Color Viewer Slider Example Update (slider-1)

jongund opened the pull request:
This is an updated version of the slider 1 example for picking a color.  

New features include:
1. Javascript file that uses a single prototype object for the example.
2. Use SVG graphics for buttons, thumbs and rail.
3. Improve high contrast support through use of  inline SVG graphics.


### Preview Link

[Color Viewer Slider Example](https://raw.githack.com/w3c/aria-practices/update-slider-1/examples/slider/slider-color-viewer.html)


### Review checklist

* [ ] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @a11ydoer 
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @a11ydoer 
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @charmarkk
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @spectranaut
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @spectranaut

spectranaut reviewed (changes requested):
code review: I only have SUPER minor comments on the code review, this looks really good! But please do delete the old javascript and css files (slide.js and slider.css) in this PR, and consider the other comments. This is otherwise an approving code review.

test review: The tests as they are look really good, but it does seems like there is a quite a bit of mouse-user functionality that is not tested. We aren't requiring tests of mouse behavior that in the test review right now, but maybe we should make a backlog task for this to not loose track of untested functionality for these widgets?

jongund commented on examples/slider/slider-1.html:
```diff
@@ -2,7 +2,7 @@
 <head>
   <meta charset="utf-8" />
-  <title>Horizontal Slider Example | WAI-ARIA Authoring Practices 1.2</title>
+  <title>Color Picker Slider Example Using SVG | WAI-ARIA Authoring Practices 1.1</title>
```
fixed

jongund commented on examples/slider/js/slider-1.js:
```diff
@@ -0,0 +1,360 @@
+*   This content is licensed according to the W3C Software License at
+*   https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document
+*
+*   File:   slider.js
```
fixed

jongund commented on examples/slider/js/slider-1.js:
```diff
@@ -0,0 +1,360 @@
+
+ColorPickerSliders.prototype.moveSliderTo = function (slider, value) {
+
+  console.log('[moveSliderTo][slider]: ' + slider.sliderNode.classList);
```
fixed

jongund commented:
@spectranaut 
Valerie, thank you for the feedback on the code, I have updated the example with your suggested changes.

spectranaut reviewed (approved):
Looks great! :)

jnurthen commented on examples/slider/css/slider-1.css:
```diff
@@ -0,0 +1,143 @@
+  text-align: center;
+  padding: 0.25em;
+}
+
```
Suggest adding 
.color-box {
  forced-color-adjust: none;
}
so the color box renders the actual color even in high contrast mode.

jnurthen commented on examples/slider/slider-1.html:
```diff
@@ -25,81 +25,209 @@
+              <h2 id="id-color-picker">Color Viewer</h2>
+              <div id="id-red" class="label">Red</div>
+              <div class="color-group red">
+                <button class="change dec10" aria-label="decrement red ten" tabindex="-1">
```
Would aria-label be better as "decrease red by 10"? (and similar for all the others?)

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+            <td><code>div</code></td>
+            <td>
+              <ul>
+                <li>Contains the ID that reference the <code>div</code> element that provide the accessible name for the group.</li>
+                <li>One ID refers to the visible label <q>Color Viewer</q>.</li>
```
The 2nd point is confusing. Suggest merging it into the first:

```suggestion
                <li>Contains the ID that refers to the <code>h2</code> element that provides the accessible name for the <q>Color Viewer</q> group.</li>
```

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+            <td></td>
+            <th scope="row"><code>aria-label="NAME_STRING"</code></th>
+            <td><code>button</code></td>
+            <td>Defines the accessible name for each increase and decrease button (<q>-10</q>, <q>-1</q>, <q>1</q>, and <q>+10</q>).</td>
```
Missing + sign.

```suggestion
            <td>Defines the accessible name for each increase and decrease button (<q>-10</q>, <q>-1</q>, <q>+1</q>, and <q>+10</q>).</td>
```

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+      <h2>Accessibility Features</h2>
+      <ul>
+        <li>
+          The three sliders are wrapped in a group to help assistive technology users understand that all three elements together represent a color picker.
```
Did you want to say "viewer" here?

```suggestion
          The three sliders are wrapped in a group to help assistive technology users understand that all three elements together represent a color viewer.
```

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+        <li>
+          The three sliders are wrapped in a group to help assistive technology users understand that all three elements together represent a color picker.
+        </li>
+        <li>Use of SVG enables high contrast support by including additional background <code>rects</code> and <code>circle</code> elements to provide contrast for the focus and hover border styling of the buttons and sliders.</li>
```
rect and circle elements (instead of rects)

```suggestion
        <li>Use of SVG enables high contrast support by including additional background <code>rect</code> and <code>circle</code> elements to provide contrast for the focus and hover border styling of the buttons and sliders.</li>
```

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+          The three sliders are wrapped in a group to help assistive technology users understand that all three elements together represent a color picker.
+        </li>
+        <li>Use of SVG enables high contrast support by including additional background <code>rects</code> and <code>circle</code> elements to provide contrast for the focus and hover border styling of the buttons and sliders.</li>
+        <li>The placement of the slider value above the thumb supports user's with low vision by allow them to more easily see the value of the slider as they move the thumb.</li>
```
remove apostrophe

```suggestion
        <li>The placement of the slider value above the thumb supports users with low vision by allow them to more easily see the value of the slider as they move the thumb.</li>
```

carmacleod commented on examples/slider/slider-color-viewer.html:
```diff
@@ -0,0 +1,422 @@
+        <li>The use of the <code>currentColor</code> for the slider value supports the visibility of the value in operating system high contrast modes.</li>
+        <li>
+          The increase and decrease buttons are not contained within the <code>div</code> with role <code>slider</code> so they can be separately focusable by users of touch screen readers.
+          However, they are excluded from the page <kbd>Tab</kbd> sequence with <code>tabindex="-1"</code> because they are redundant with the arrow key support provided to keyboard users.
```
```suggestion
          However, they are excluded from the page <kbd>Tab</kbd> sequence with <code>tabindex="-1"</code> because they are redundant with the arrow key support provided for keyboard users.
```

carmacleod reviewed (changes requested):
Looks great, @jongund !

I just have a few "Github Suggestions", as noted above.

Most of the suggestions are editorial or labelling changes, and 2 of the suggestions will help make the Open in Codepen button work.

One additional suggestion is to please add `focusable="false"` to all of the SVGs so that they don't take focus on IE11. (I know we're no longer claiming support for IE11, but it's such a little thing that makes a big difference).

nschonni commented on test/tests/slider_slider-color-viewer.js:
```diff
@@ -423,14 +498,16 @@ ariaTest(
 
     t.is(
-      await redSlider.getAttribute('aria-valuenow'),
+      await await redSlider.getAttribute('aria-valuenow'),
```
```suggestion
      await redSlider.getAttribute('aria-valuenow'),
```

jongund replied:
@nschonni 
Had to put in two "awaits", otherwise the test intermittently failed.  Not sure why, but that fixed it for me at least.  I believe Valarie has used this technique on some other tests, that's why I tired it.

nschonni replied:
Fair enough 😄

carmacleod commented:
Please see the suggestion here: https://github.com/w3c/aria-practices/pull/1472/files#r515402891

I think the following 2 points are confusing:
- Contains the ID that reference the div element that provide the accessible name for the group.
- One ID refers to the visible label Color Viewer.

Maybe merge them into 1 point, as follows:
- Contains the ID that refers to the h2 element that provides the accessible name for the Color Viewer group.

carmacleod reviewed (approved):
Assuming my last [comment](https://github.com/w3c/aria-practices/pull/1472#issuecomment-736800784) is addressed, I approve!
Thanks, @jongund!

jongund commented:
Created separate Color Picker example without the increment and decrement buttons.

jongund commented:
@spectranaut @nschonni @mcking65 
One of the checks for color viewer example is failing because it said I modified "examples\index.html", you can see from the files changed that I am not modifying "examples/index.html".  How does this get fixed?

nschonni commented:
Run `npm run reference-tables` to update the examples index.html with the new file names

jongund commented:
> 
> 
> Run `npm run reference-tables` to update the examples index.html with the new file names

@nschonni 
Thank you for quick response.
I tried running the command and it did say in "generating index" but it didn't change the examples/index.html file (or anything else as far as I can tell), so no changes to commit.

jongund commented:
@nschonni 
Looks like the problem fixed itself, it's not reporting a code issue anymore.

nschonni commented:
@jongund there is also a pre-commit hook now, so it looked like it regenerated itself in https://github.com/w3c/aria-practices/pull/1472/commits/403645ea13a5b4d666b49f107ec39dd7213057f5

charmarkk reviewed (approved):
Tested with Windows 10, NVDA 2020.3, JAWS 2020.2012.13; Firefox 84.0.1 and Chrome 87.0.4280.141. All worked as expected as far as documented accessibility functionality, keyboard interaction, Windows HCM, and screen readers were concerned. 

@carmacleod, could I tag you in for MacOS a11y tests? My test Mac is with another co-worker at the moment.

a11ydoer commented:
Since @charmarkk did functionality review agaist on non iOS devices, I only check color picker slider example with mobile support. 

The slider itself works well by annoucing current value or changed value with gesture (See the screen shot with VoiceOver text output). 

Question is on increase/decrease button. For example, when "increase red by ten" is touched with gesture, the value changed  accordingly.  (See the screen shot with VoiceOver text output). However, the changed value itself is not announced although visual user can see the color change in the Color view box. @jongund Should we announce changed value with button action or it is not necessary?

a11ydoer commented:
![focus on increase red by ten but it does not annonce the chaged value ](https://user-images.githubusercontent.com/14183123/104239690-8ca14d00-5420-11eb-8d2c-6761f517e6c5.PNG)
![focus on red slider - announce the value correctly with gesture](https://user-images.githubusercontent.com/14183123/104239692-8d39e380-5420-11eb-931f-8fa579a4d8cf.PNG)

a11ydoer commented:
Regarding visual design review, we do not have much option since APG examples are not responsive yet. Furthermore, other color picker design examples tend to be 3 dimensional which is complex to implement in terms of accessiblity. (See the screen shot of "color zilla" example, the circle with RGB color combination) Therefore, current color picker implementation with 2D linear slider design is acceptable. 
![color picker tool with 3d circle RGB color](https://user-images.githubusercontent.com/14183123/104240390-9bd4ca80-5421-11eb-8374-3d52f149d90a.png)

jongund commented:
@nschonni @mcking65 
The same problem as the other pull request I had last week, it is finding a problem in the `examples/index.html` file that I did not modify in the in this pull request.

nschonni commented:
You deleted files so the index.html is out out of sync. Last time it got regenerated by the pre-commit hook during an unrelated changed, but you can run the script from the previous comment to do it

jongund commented:
@mcking65 
I have removed the color viewer mobile example from this pull request and updated the accessibility documentation.   So it is ready for editorial review.

jongund commented:
This pull request is superseded by pull #1746.
